### PR TITLE
adapt vllm distributed module to sglang

### PR DESCRIPTION
**Pull request by yizhang2077**
## Modifications

Move vllm distributed module (v0.6.4.post1) to sglang, and current model is stilll using vllm.distributed module

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

## What to do Next

- [ ] Install sgl-kernel for custom allreduce and replace vllm.distributed to sglang.srt.distributed
- [ ] Add tensorrt allreduce module

### CODE REVIEW

**Comment by zhyncs** [comment]
QQ why not use the v0.6.4.post1

**Comment by yizhang2077** [comment]
> QQ why not use the v0.6.4.post1

I see pyproject.toml require vllm>=0.6.3.post1

**Comment by zhyncs** [comment]
This is to maintain compatibility, 0.6.3.post1 uses torch 2.4, and 0.6.4.post1 uses torch 2.5.1. The current main branch is compatible with both torch 2.4 and torch 2.5.1. Regarding the distributed part, I suggest directly updating to v0.6.4.post1

**Comment by yizhang2077** [comment]
> This is to maintain compatibility, 0.6.3.post1 uses torch 2.4, and 0.6.4.post1 uses torch 2.5.1. The current main branch is compatible with both torch 2.4 and torch 2.5.1. Regarding the distributed part, I suggest directly updating to v0.6.4.post1

ok，currently it use 0.6.4.post1 as base

File: python/sglang/srt/_custom_ops.py [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+import functools
+import importlib
+import logging
+from typing import TYPE_CHECKING, List, Optional, Tuple, Union
+
+import torch
+import torch.library
+
+# import vllm.envs as envs
+from vllm.platforms import current_platform
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py]: May we remove `current_platform`

File: python/sglang/srt/_custom_ops.py [hunk 2]
```diff
@@ -0,0 +1,1005 @@
...
+logger = logging.getLogger(__name__)
+
+# if not current_platform.is_tpu() and not current_platform.is_hpu():
+if not current_platform.is_tpu() and not is_hpu():
+    try:
+        import custom_ar
+    except ImportError as e:
+        logger.warning("Failed to import from custom_ar with %r", e)
+
+"""
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py]: Please remove all the unused code

File: python/sglang/srt/_custom_ops.py [hunk 3]
```diff
@@ -0,0 +1,1005 @@
...
+                "Please try a clean build and install of vllm,"
+                "or remove old built files such as vllm/*cpython*.so and build/ ."
+            )
+            logger.error(msg, fn.__name__, e)
+            raise e
+
+    return wrapper
+
+
+'''
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py]: Same as above

File: python/sglang/srt/_custom_ops.py [hunk 4]
```diff
@@ -0,0 +1,1005 @@
+import contextlib
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py:2]: Please add source (e.g. commit id or version from vllm)

File: python/sglang/srt/distributed/device_communicators/cuda_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
+"""This file is a pure Python wrapper for the cudart library.
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/cuda_wrapper.py:2]: Please add the vllm source version here

File: python/sglang/srt/distributed/device_communicators/custom_all_reduce.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
+import ctypes
+import logging
+from contextlib import contextmanager
+from typing import List, Optional, Union
+
+import torch
+import torch.distributed as dist
+import vllm.envs as envs
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/custom_all_reduce.py]: `envs`, `current_platform` and `cuda_device_count_stateless` should also be replaced

File: python/sglang/srt/distributed/device_communicators/custom_all_reduce.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
+import ctypes
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/custom_all_reduce.py:2]: Please add the vllm source version here

File: python/sglang/srt/distributed/device_communicators/custom_all_reduce_utils.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+import pickle
+import subprocess
+import sys
+import tempfile
+from itertools import product
+from typing import Dict, List, Optional, Sequence
+
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import vllm.envs as envs
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/custom_all_reduce_utils.py]: Same as above

File: python/sglang/srt/distributed/device_communicators/custom_all_reduce_utils.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
+import ctypes
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/custom_all_reduce_utils.py:2]: Please add the vllm source version here

File: python/sglang/srt/distributed/device_communicators/pynccl.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
+# ===================== import region =====================
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/pynccl.py]: Please add the vllm source version here and remove unused comment

File: python/sglang/srt/distributed/device_communicators/pynccl_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+
+import ctypes
+import logging
+import platform
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional
+
+import torch
+from torch.distributed import ReduceOp
+from vllm.utils import find_nccl_library
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/pynccl_wrapper.py]: Please replace `find_nccl_library`

File: python/sglang/srt/distributed/device_communicators/tpu_communicator.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import os
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/tpu_communicator.py]: `tpu_communicator` is useless for SGLang for now and can be considered not to introduce it

File: python/sglang/srt/distributed/device_communicators/shm_broadcast.py [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from multiprocessing import shared_memory
+from typing import List, Optional
+from unittest.mock import patch
+
+import torch
+import torch.distributed as dist
+import vllm.envs as envs
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/shm_broadcast.py]: Please remove `envs` and `vllm.utils`

File: python/sglang/srt/distributed/device_communicators/pynccl_wrapper.py [hunk 14]
```diff
@@ -0,0 +1,332 @@
+# This file is a pure Python wrapper for the NCCL library.
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/pynccl_wrapper.py:3]: Please add the vllm source version here

File: python/sglang/srt/distributed/device_communicators/shm_broadcast.py [hunk 15]
```diff
@@ -0,0 +1,496 @@
+import logging
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/shm_broadcast.py:3]: Please add the vllm source version here

File: python/sglang/srt/distributed/parallel_state.py [hunk 16]
```diff
@@ -0,0 +1,1291 @@
+# Copyright 2023 The vLLM team.
```
Comment (zhyncs) [inline python/sglang/srt/distributed/parallel_state.py:3]: Please add the vllm source version here

File: python/sglang/srt/distributed/device_communicators/xpu_communicator.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
+import torch
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+from vllm.platforms import current_platform
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/xpu_communicator.py]: Please replace `current_platform`

**Review by zhyncs — COMMENTED** [review]
Overall LGTM just left some comments Thanks!

File: python/sglang/srt/distributed/communication_op.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
+# reference: VLLM 0.6.4.post1
```
Comment (zhyncs) [inline python/sglang/srt/distributed/communication_op.py]:
```suggestion
# Adapted from https://github.com/vllm-project/vllm/blob/a6221a144af772fd1a68fe7e627935dc53e81738/vllm/distributed/communication_op.py
```

Comment (zhyncs) [reply python/sglang/srt/distributed/communication_op.py]: something like this

File: python/sglang/srt/_custom_ops.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+import importlib
+import logging
+from typing import TYPE_CHECKING, List, Optional, Tuple, Union
+
+import torch
+import torch.library
+
+from sglang.srt.utils import is_hpu
+
+# from vllm.scalar_type import ScalarType
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py]:
```suggestion
```

File: python/sglang/srt/distributed/device_communicators/custom_all_reduce.py [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+                        "NVLink detection failed. This is normal if your"
+                        " machine has no NVLink equipped."
+                    )
+                    return False
+    return True
+
+
+def _can_p2p(rank: int, world_size: int) -> bool:
+    # SGLANG_SKIP_P2P_CHECK can be set to False in sglang
+    SGLANG_SKIP_P2P_CHECK = os.getenv("VLLM_SKIP_P2P_CHECK", "0") == "1"
```
Comment (zhyncs) [inline python/sglang/srt/distributed/device_communicators/custom_all_reduce.py]: May you replace all `VLLM_` with `SGLANG_`

File: python/sglang/srt/_custom_ops.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
+# Adapted from https://github.com/vllm-project/vllm/blob/a6221a144af772fd1a68fe7e627935dc53e81738/vllm/_custom_ops.py
```
Comment (zhyncs) [inline python/sglang/srt/_custom_ops.py:1]: BTW use pre-commit or `make format` to fix the lint issue

**Review by zhyncs — COMMENTED** [review]
LGTM. I think we can merge it into `remove-vllm-distributed` and verify afterward. You can then create another PR from `remove-vllm-distributed`.

**Comment by zhyncs** [comment]
```
isort....................................................................Failed
- hook id: isort
- files were modified by this hook

Fixing /home/runner/work/sglang/sglang/python/sglang/srt/_custom_ops.py
```

**Comment by yizhang2077** [comment]
I think the test failed is caused by pr #2266 , @zhaochenyang20

**Review by zhyncs — APPROVED** [review]
LGTM Since it is not currently in use, I think it's safe to merge. More detailed testing and verification can be conducted when integrating the custom all-reduce CUDA kernel into sgl-kernel. cc @yizhang2077

File: python/sglang/srt/distributed/__init__.py [hunk 22]
```diff
@@ -0,0 +1,3 @@
+from .communication_op import *
```
Comment (merrymercy) [inline python/sglang/srt/distributed/__init__.py:1]: Use absolute path.

File: python/sglang/srt/distributed/communication_op.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
+# Adapted from https://github.com/vllm-project/vllm/blob/a6221a144af772fd1a68fe7e627935dc53e81738/vllm/distributed/communication_op.py
+from typing import Any, Dict, Optional, Union
+
+import torch
+import torch.distributed
+
+from .parallel_state import get_tp_group
```
Comment (merrymercy) [inline python/sglang/srt/distributed/communication_op.py:7]: use absolute path.

Comment (yizhang2077) [reply python/sglang/srt/distributed/communication_op.py:7]: OK，I will fix it when I add custom_ar to sgl-kernel

File: python/sglang/srt/_custom_ops.py [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (merrymercy) [inline python/sglang/srt/_custom_ops.py:1]: If this is exactly a version release commit, can you use something like
https://github.com/vllm-project/vllm/blob/v0.6.4.post1/vllm/_custom_ops.py

**Pull request merged**